onc-healthit/lantern-back-end
# Backup and Restore Database

Changes in this PR:
This PR now allows you to create database backup files that can simply be used to restore the database without needing to reload and reprocess all the information to populate the db, which takes a lot of time. The make command to make a backup file for the database is `make backup_database` and the command to restore the database is `make restore_database file=<backup file name>` Backup files that are created are named "lantern_backup_`date +%Y%m%d%H%M%S`.sql" (where +%Y%m%d%H%M%S is the timestamp for when the file was created) and they can be found in the base directory after running the make backup_database command.

TO TEST:
Run the lantern project with the make run command, and populate the db using make populatedb. Check the database to make sure it is populated. Then, run make backup_database. Next, stop lantern from running and run make clean to clean the database. Run the lantern project again with make run, and check that the database is now empty. Now, run make restore_database file=<name of the backup file you just created> and check the database again. It now should be populated exactly as it was after running make popualtedb.

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: 
[Lantern 96 Link](https://oncprojectracking.healthit.gov/support/browse/LANTERN-96)
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
